File: pqc_abstract.py
"""
pqc_abstract.py — Post-Quantum Cryptography Abstract Layer
===========================================================
A clean, Pythonic wrapper around the ``oqs`` (liboqs-python) package that
exposes KEM and SIG primitives through safe, self-documenting APIs.

Dependencies
------------
    pip install liboqs-python        # installs the ``oqs`` package
    # liboqs shared library must also be built and installed system-wide.
    # See: https://github.com/open-quantum-safe/liboqs

Quick start
-----------
    from pqc_abstract import KEM, SIG, PQCError

    # --- KEM round-trip ---
    kem = KEM()                          # defaults to ML-KEM-768
    alice = kem.keygen()
    enc   = kem.encaps(alice.public_key)
    ss    = kem.decaps(enc.ciphertext, alice.secret_key)
    assert enc.shared_secret == ss

    # --- SIG round-trip ---
    sig   = SIG()                        # defaults to ML-DSA-65
    keys  = sig.keygen()
    s     = sig.sign(b"hello world", keys.secret_key)
    ok    = sig.verify(b"hello world", s, keys.public_key)

SPDX-License-Identifier: MIT
"""

from __future__ import annotations

import importlib
import secrets
from dataclasses import dataclass, field
from typing import Optional

# ---------------------------------------------------------------------------
# Lazy import of the ``oqs`` package so that import of *this* module does not
# crash in environments where liboqs is not installed.
# ---------------------------------------------------------------------------
def _require_oqs():
    try:
        import oqs  # noqa: F401 – checked below
    except ImportError as exc:
        raise ImportError(
            "The 'oqs' package (liboqs-python) is required. "
            "Install with:  pip install liboqs-python"
        ) from exc
    return importlib.import_module("oqs")


# ═══════════════════════════════════════════════════════════════════════════
# Exceptions
# ═══════════════════════════════════════════════════════════════════════════

class PQCError(Exception):
    """Base class for all pqc_abstract errors."""


class AlgorithmNotAvailable(PQCError):
    """Raised when a requested algorithm is unknown or disabled at build time."""


class KeygenError(PQCError):
    """Raised when key generation fails."""


class EncapsulationError(PQCError):
    """Raised when KEM encapsulation fails."""


class DecapsulationError(PQCError):
    """Raised when KEM decapsulation fails."""


class SigningError(PQCError):
    """Raised when signing fails."""


class VerificationError(PQCError):
    """Raised when signature verification fails (invalid signature)."""


# ═══════════════════════════════════════════════════════════════════════════
# Well-known algorithm names
# ═══════════════════════════════════════════════════════════════════════════

class KEMAlgorithm:
    """String constants for KEM algorithm names."""
    DEFAULT        = "ML-KEM-768"

    ML_KEM_512     = "ML-KEM-512"
    ML_KEM_768     = "ML-KEM-768"
    ML_KEM_1024    = "ML-KEM-1024"

    KYBER_512      = "Kyber512"
    KYBER_768      = "Kyber768"
    KYBER_1024     = "Kyber1024"

    BIKE_L1        = "BIKE-L1"
    BIKE_L3        = "BIKE-L3"
    BIKE_L5        = "BIKE-L5"

    FRODO_640_AES  = "FrodoKEM-640-AES"
    FRODO_640_SHAKE= "FrodoKEM-640-SHAKE"
    FRODO_976_AES  = "FrodoKEM-976-AES"
    FRODO_976_SHAKE= "FrodoKEM-976-SHAKE"

    HQC_1          = "HQC-1"
    HQC_3          = "HQC-3"
    HQC_5          = "HQC-5"


class SIGAlgorithm:
    """String constants for signature algorithm names."""
    DEFAULT        = "ML-DSA-65"

    ML_DSA_44      = "ML-DSA-44"
    ML_DSA_65      = "ML-DSA-65"
    ML_DSA_87      = "ML-DSA-87"

    FALCON_512     = "Falcon-512"
    FALCON_1024    = "Falcon-1024"
    FALCON_PADDED_512  = "Falcon-padded-512"
    FALCON_PADDED_1024 = "Falcon-padded-1024"

    MAYO_1         = "MAYO-1"
    MAYO_2         = "MAYO-2"
    MAYO_3         = "MAYO-3"
    MAYO_5         = "MAYO-5"


# ═══════════════════════════════════════════════════════════════════════════
# Data classes — returned by API methods
# ═══════════════════════════════════════════════════════════════════════════

@dataclass
class KEMKeyPair:
    """Holds a KEM public/secret key pair."""
    public_key: bytes
    secret_key: bytes

    def __repr__(self) -> str:
        return (
            f"KEMKeyPair(pk={self.public_key[:8].hex()}… [{len(self.public_key)} B], "
            f"sk=<{len(self.secret_key)} B hidden>)"
        )

    def wipe(self) -> None:
        """Best-effort in-memory zeroise (note: Python GC may keep copies)."""
        object.__setattr__(self, "secret_key", bytes(len(self.secret_key)))


@dataclass
class KEMEncapsResult:
    """Result of a KEM encapsulation operation."""
    ciphertext:    bytes
    shared_secret: bytes

    def __repr__(self) -> str:
        return (
            f"KEMEncapsResult(ct={self.ciphertext[:8].hex()}… [{len(self.ciphertext)} B], "
            f"ss={self.shared_secret[:8].hex()}… [{len(self.shared_secret)} B])"
        )

    def wipe(self) -> None:
        object.__setattr__(self, "shared_secret", bytes(len(self.shared_secret)))


@dataclass
class SIGKeyPair:
    """Holds a SIG public/secret key pair."""
    public_key: bytes
    secret_key: bytes

    def __repr__(self) -> str:
        return (
            f"SIGKeyPair(pk={self.public_key[:8].hex()}… [{len(self.public_key)} B], "
            f"sk=<{len(self.secret_key)} B hidden>)"
        )

    def wipe(self) -> None:
        object.__setattr__(self, "secret_key", bytes(len(self.secret_key)))


@dataclass
class AlgorithmInfo:
    """Algorithm metadata."""
    name:               str
    nist_level:         int
    public_key_length:  int
    secret_key_length:  int
    # KEM specific
    ciphertext_length:  Optional[int] = None
    shared_secret_length: Optional[int] = None
    ind_cca:            Optional[bool] = None
    # SIG specific
    max_signature_length: Optional[int] = None
    euf_cma:            Optional[bool] = None
    suf_cma:            Optional[bool] = None
    ctx_str_support:    Optional[bool] = None

    def __str__(self) -> str:
        lines = [f"Algorithm : {self.name}", f"  NIST level : {self.nist_level}"]
        lines.append(f"  Public key : {self.public_key_length} B")
        lines.append(f"  Secret key : {self.secret_key_length} B")
        if self.ciphertext_length is not None:
            lines.append(f"  Ciphertext : {self.ciphertext_length} B")
            lines.append(f"  Shared secret : {self.shared_secret_length} B")
            lines.append(f"  IND-CCA2 : {self.ind_cca}")
        if self.max_signature_length is not None:
            lines.append(f"  Max signature : {self.max_signature_length} B")
            lines.append(f"  EUF-CMA : {self.euf_cma}")
            lines.append(f"  SUF-CMA : {self.suf_cma}")
            lines.append(f"  Context strings : {self.ctx_str_support}")
        return "\n".join(lines)


# ═══════════════════════════════════════════════════════════════════════════
# KEM
# ═══════════════════════════════════════════════════════════════════════════

class KEM:
    """
    Post-Quantum Key Encapsulation Mechanism.

    Parameters
    ----------
    algorithm : str, optional
        liboqs algorithm name.  Defaults to ``KEMAlgorithm.DEFAULT``
        ("ML-KEM-768", NIST FIPS 203 level 3).

    Examples
    --------
    >>> kem   = KEM("ML-KEM-512")
    >>> keys  = kem.keygen()
    >>> enc   = kem.encaps(keys.public_key)
    >>> ss    = kem.decaps(enc.ciphertext, keys.secret_key)
    >>> assert enc.shared_secret == ss
    """

    def __init__(self, algorithm: str = KEMAlgorithm.DEFAULT) -> None:
        oqs = _require_oqs()
        self._algorithm = algorithm
        try:
            self._kem = oqs.KeyEncapsulation(algorithm)
        except Exception as exc:
            raise AlgorithmNotAvailable(
                f"KEM algorithm '{algorithm}' is not available. "
                "Check that liboqs was built with that algorithm enabled."
            ) from exc

    # ── Properties ──────────────────────────────────────────────────────────

    @property
    def algorithm(self) -> str:
        return self._algorithm

    @property
    def info(self) -> AlgorithmInfo:
        d = self._kem.details
        return AlgorithmInfo(
            name=d["name"],
            nist_level=d["claimed_nist_level"],
            public_key_length=d["length_public_key"],
            secret_key_length=d["length_secret_key"],
            ciphertext_length=d["length_ciphertext"],
            shared_secret_length=d["length_shared_secret"],
            ind_cca=d["ind_cca"],
        )

    # ── Core API ─────────────────────────────────────────────────────────────

    def keygen(self) -> KEMKeyPair:
        """Generate a fresh KEM key pair.

        Returns
        -------
        KEMKeyPair
            Freshly generated public/secret key pair.

        Raises
        ------
        KeygenError
        """
        try:
            public_key = self._kem.generate_keypair()
            secret_key = self._kem.export_secret_key()
            return KEMKeyPair(public_key=bytes(public_key),
                              secret_key=bytes(secret_key))
        except Exception as exc:
            raise KeygenError(f"KEM keygen failed: {exc}") from exc

    def encaps(self, public_key: bytes) -> KEMEncapsResult:
        """Encapsulate: generate a ciphertext and shared secret.

        Parameters
        ----------
        public_key : bytes
            Recipient's public key (``info.public_key_length`` bytes).

        Returns
        -------
        KEMEncapsResult
            Contains ``ciphertext`` and ``shared_secret``.

        Raises
        ------
        EncapsulationError
        """
        try:
            # liboqs-python encap_secret returns (ciphertext, shared_secret)
            ciphertext, shared_secret = self._kem.encap_secret(public_key)
            return KEMEncapsResult(ciphertext=bytes(ciphertext),
                                   shared_secret=bytes(shared_secret))
        except Exception as exc:
            raise EncapsulationError(f"KEM encapsulation failed: {exc}") from exc

    def decaps(self, ciphertext: bytes, secret_key: bytes) -> bytes:
        """Decapsulate: recover the shared secret.

        Parameters
        ----------
        ciphertext : bytes
            Ciphertext from :meth:`encaps`.
        secret_key : bytes
            Recipient's secret key.

        Returns
        -------
        bytes
            Recovered shared secret (``info.shared_secret_length`` bytes).

        Raises
        ------
        DecapsulationError
        """
        try:
            # We need a fresh KEM object seeded with the secret key.
            oqs = _require_oqs()
            kem_dec = oqs.KeyEncapsulation(self._algorithm, secret_key)
            shared_secret = kem_dec.decap_secret(ciphertext)
            return bytes(shared_secret)
        except Exception as exc:
            raise DecapsulationError(f"KEM decapsulation failed: {exc}") from exc

    def __repr__(self) -> str:
        return f"KEM(algorithm={self._algorithm!r})"

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self._kem.free()


# ═══════════════════════════════════════════════════════════════════════════
# SIG
# ═══════════════════════════════════════════════════════════════════════════

class SIG:
    """
    Post-Quantum Digital Signature Scheme.

    Parameters
    ----------
    algorithm : str, optional
        liboqs algorithm name.  Defaults to ``SIGAlgorithm.DEFAULT``
        ("ML-DSA-65", NIST FIPS 204 level 3).

    Examples
    --------
    >>> sig  = SIG("Falcon-512")
    >>> keys = sig.keygen()
    >>> s    = sig.sign(b"hello", keys.secret_key)
    >>> assert sig.verify(b"hello", s, keys.public_key)
    """

    def __init__(self, algorithm: str = SIGAlgorithm.DEFAULT) -> None:
        oqs = _require_oqs()
        self._algorithm = algorithm
        try:
            self._sig = oqs.Signature(algorithm)
        except Exception as exc:
            raise AlgorithmNotAvailable(
                f"SIG algorithm '{algorithm}' is not available. "
                "Check that liboqs was built with that algorithm enabled."
            ) from exc

    # ── Properties ──────────────────────────────────────────────────────────

    @property
    def algorithm(self) -> str:
        return self._algorithm

    @property
    def info(self) -> AlgorithmInfo:
        d = self._sig.details
        return AlgorithmInfo(
            name=d["name"],
            nist_level=d["claimed_nist_level"],
            public_key_length=d["length_public_key"],
            secret_key_length=d["length_secret_key"],
            max_signature_length=d["length_signature"],
            euf_cma=d["euf_cma"],
            suf_cma=d.get("suf_cma", False),
            ctx_str_support=d.get("sig_with_ctx_support", False),
        )

    # ── Core API ─────────────────────────────────────────────────────────────

    def keygen(self) -> SIGKeyPair:
        """Generate a fresh SIG key pair.

        Returns
        -------
        SIGKeyPair

        Raises
        ------
        KeygenError
        """
        try:
            public_key = self._sig.generate_keypair()
            secret_key = self._sig.export_secret_key()
            return SIGKeyPair(public_key=bytes(public_key),
                              secret_key=bytes(secret_key))
        except Exception as exc:
            raise KeygenError(f"SIG keygen failed: {exc}") from exc

    def sign(self,
             message: bytes,
             secret_key: bytes,
             *,
             ctx: Optional[bytes] = None) -> bytes:
        """Sign a message.

        Parameters
        ----------
        message : bytes
            Arbitrary message bytes.
        secret_key : bytes
            Signer's secret key.
        ctx : bytes, optional
            Context string (only for algorithms that support it,
            e.g. ML-DSA).  Silently ignored if the algorithm does
            not support context strings.

        Returns
        -------
        bytes
            The signature.

        Raises
        ------
        SigningError
        """
        try:
            oqs = _require_oqs()
            sig_obj = oqs.Signature(self._algorithm, secret_key)
            if ctx is not None and self.info.ctx_str_support:
                signature = sig_obj.sign_with_ctx_str(message, ctx)
            else:
                signature = sig_obj.sign(message)
            return bytes(signature)
        except Exception as exc:
            raise SigningError(f"Signing failed: {exc}") from exc

    def verify(self,
               message: bytes,
               signature: bytes,
               public_key: bytes,
               *,
               ctx: Optional[bytes] = None) -> bool:
        """Verify a signature.

        Parameters
        ----------
        message : bytes
            The original message.
        signature : bytes
            Signature produced by :meth:`sign`.
        public_key : bytes
            Signer's public key.
        ctx : bytes, optional
            Context string used during signing (must match).

        Returns
        -------
        bool
            ``True`` if the signature is valid.

        Raises
        ------
        VerificationError
            If the signature is definitively invalid (rather than just
            returning ``False``).
        """
        try:
            if ctx is not None and self.info.ctx_str_support:
                return bool(self._sig.verify_with_ctx_str(message, signature, ctx, public_key))
            return bool(self._sig.verify(message, signature, public_key))
        except Exception as exc:
            raise VerificationError(f"Verification error: {exc}") from exc

    def __repr__(self) -> str:
        return f"SIG(algorithm={self._algorithm!r})"

    def __enter__(self):
        return self

    def __exit__(self, *_):
        self._sig.free()


# ═══════════════════════════════════════════════════════════════════════════
# Utility helpers
# ═══════════════════════════════════════════════════════════════════════════

def list_kem_algorithms(*, enabled_only: bool = True) -> list[str]:
    """Return names of available KEM algorithms.

    Parameters
    ----------
    enabled_only : bool
        If True (default), skip algorithms disabled at compile time.
    """
    oqs = _require_oqs()
    algs = oqs.get_enabled_kem_mechanisms() if enabled_only else oqs.get_supported_kem_mechanisms()
    return list(algs)


def list_sig_algorithms(*, enabled_only: bool = True) -> list[str]:
    """Return names of available SIG algorithms."""
    oqs = _require_oqs()
    algs = oqs.get_enabled_sig_mechanisms() if enabled_only else oqs.get_supported_sig_mechanisms()
    return list(algs)


def kem_info(algorithm: str) -> AlgorithmInfo:
    """Return metadata for a KEM algorithm without creating a full context."""
    return KEM(algorithm).info


def sig_info(algorithm: str) -> AlgorithmInfo:
    """Return metadata for a SIG algorithm without creating a full context."""
    return SIG(algorithm).info

File: pqc_abstract.c
/**
 * @file pqc_abstract.c
 * @brief Implementation of the Post-Quantum Cryptography Abstract Layer.
 *
 * SPDX-License-Identifier: MIT
 */

#include "pqc_abstract.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

/* ── liboqs public headers ─────────────────────────────────────────────────── */
#include <oqs/oqs.h>
#include <oqs/kem.h>
#include <oqs/sig.h>

/* ═══════════════════════════════════════════════════════════════════════════
 * Internal helpers
 * ═══════════════════════════════════════════════════════════════════════════ */

/** Zeroise and free a heap buffer. */
static void _secure_free(uint8_t **buf, size_t len) {
    if (buf && *buf) {
        OQS_MEM_cleanse(*buf, len);
        OQS_MEM_secure_free(*buf, len);
        *buf = NULL;
    }
}

/* ═══════════════════════════════════════════════════════════════════════════
 * Library lifecycle
 * ═══════════════════════════════════════════════════════════════════════════ */

void pqc_init(void) {
    OQS_init();
}

void pqc_destroy(void) {
    OQS_destroy();
}

/* ═══════════════════════════════════════════════════════════════════════════
 * Status helpers
 * ═══════════════════════════════════════════════════════════════════════════ */

const char *pqc_status_str(PQC_STATUS s) {
    switch (s) {
        case PQC_OK:              return "OK";
        case PQC_ERR_GENERIC:     return "Generic error";
        case PQC_ERR_NULL:        return "NULL argument or uninitialised context";
        case PQC_ERR_ALG_UNKNOWN: return "Unknown algorithm name";
        case PQC_ERR_ALG_DISABLED:return "Algorithm disabled at compile time";
        case PQC_ERR_KEYGEN:      return "Key-generation failure";
        case PQC_ERR_ENCAPS:      return "Encapsulation failure";
        case PQC_ERR_DECAPS:      return "Decapsulation failure";
        case PQC_ERR_SIGN:        return "Signing failure";
        case PQC_ERR_VERIFY:      return "Signature verification failed";
        case PQC_ERR_ALLOC:       return "Memory allocation failure";
        default:                   return "Unknown status code";
    }
}

/* ═══════════════════════════════════════════════════════════════════════════
 * KEM — context management
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_KEM_Ctx *pqc_kem_new(const char *alg) {
    const char *name = alg ? alg : PQC_KEM_DEFAULT;

    OQS_KEM *kem = OQS_KEM_new(name);
    if (!kem) {
        /* Distinguish unknown vs disabled */
        fprintf(stderr, "[pqc] KEM '%s': not available (unknown or disabled at build time)\n", name);
        return NULL;
    }

    PQC_KEM_Ctx *ctx = (PQC_KEM_Ctx *)calloc(1, sizeof(PQC_KEM_Ctx));
    if (!ctx) {
        OQS_KEM_free(kem);
        return NULL;
    }

    ctx->_kem                  = kem;
    ctx->algorithm             = kem->method_name;
    ctx->length_public_key     = kem->length_public_key;
    ctx->length_secret_key     = kem->length_secret_key;
    ctx->length_ciphertext     = kem->length_ciphertext;
    ctx->length_shared_secret  = kem->length_shared_secret;
    ctx->nist_level            = (uint8_t)kem->claimed_nist_level;
    ctx->ind_cca               = kem->ind_cca;
    return ctx;
}

void pqc_kem_free(PQC_KEM_Ctx *ctx) {
    if (!ctx) return;
    OQS_KEM_free(ctx->_kem);
    ctx->_kem = NULL;
    free(ctx);
}

/* ═══════════════════════════════════════════════════════════════════════════
 * KEM — key-pair operations
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_STATUS pqc_kem_keygen(const PQC_KEM_Ctx *ctx, PQC_KEM_Keys *keys) {
    if (!ctx || !ctx->_kem || !keys) return PQC_ERR_NULL;

    memset(keys, 0, sizeof(*keys));

    keys->pk_len = ctx->length_public_key;
    keys->sk_len = ctx->length_secret_key;

    keys->public_key = (uint8_t *)malloc(keys->pk_len);
    keys->secret_key = (uint8_t *)malloc(keys->sk_len);
    if (!keys->public_key || !keys->secret_key) {
        pqc_kem_free_keys(keys);
        return PQC_ERR_ALLOC;
    }

    OQS_STATUS rc = OQS_KEM_keypair(ctx->_kem, keys->public_key, keys->secret_key);
    if (rc != OQS_SUCCESS) {
        pqc_kem_free_keys(keys);
        return PQC_ERR_KEYGEN;
    }
    return PQC_OK;
}

void pqc_kem_free_keys(PQC_KEM_Keys *keys) {
    if (!keys) return;
    _secure_free(&keys->public_key, keys->pk_len);
    _secure_free(&keys->secret_key, keys->sk_len);
    keys->pk_len = keys->sk_len = 0;
}

/* ═══════════════════════════════════════════════════════════════════════════
 * KEM — encapsulation / decapsulation
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_STATUS pqc_kem_encaps(const PQC_KEM_Ctx *ctx,
                           const uint8_t     *pk,
                           PQC_KEM_Encaps    *result) {
    if (!ctx || !ctx->_kem || !pk || !result) return PQC_ERR_NULL;

    memset(result, 0, sizeof(*result));
    result->ct_len = ctx->length_ciphertext;
    result->ss_len = ctx->length_shared_secret;

    result->ciphertext    = (uint8_t *)malloc(result->ct_len);
    result->shared_secret = (uint8_t *)malloc(result->ss_len);
    if (!result->ciphertext || !result->shared_secret) {
        pqc_kem_free_encaps(result);
        return PQC_ERR_ALLOC;
    }

    OQS_STATUS rc = OQS_KEM_encaps(ctx->_kem,
                                    result->ciphertext,
                                    result->shared_secret,
                                    pk);
    if (rc != OQS_SUCCESS) {
        pqc_kem_free_encaps(result);
        return PQC_ERR_ENCAPS;
    }
    return PQC_OK;
}

PQC_STATUS pqc_kem_decaps(const PQC_KEM_Ctx *ctx,
                           const uint8_t     *ciphertext,
                           const uint8_t     *sk,
                           uint8_t           *shared_secret) {
    if (!ctx || !ctx->_kem || !ciphertext || !sk || !shared_secret)
        return PQC_ERR_NULL;

    OQS_STATUS rc = OQS_KEM_decaps(ctx->_kem, shared_secret, ciphertext, sk);
    return (rc == OQS_SUCCESS) ? PQC_OK : PQC_ERR_DECAPS;
}

void pqc_kem_free_encaps(PQC_KEM_Encaps *e) {
    if (!e) return;
    _secure_free(&e->ciphertext,    e->ct_len);
    _secure_free(&e->shared_secret, e->ss_len);
    e->ct_len = e->ss_len = 0;
}

/* ═══════════════════════════════════════════════════════════════════════════
 * KEM — introspection utilities
 * ═══════════════════════════════════════════════════════════════════════════ */

void pqc_kem_list_algorithms(void) {
    printf("Available KEM algorithms:\n");
    for (size_t i = 0; i < OQS_KEM_algs_length; i++) {
        const char *name = OQS_KEM_alg_identifier(i);
        if (OQS_KEM_alg_is_enabled(name))
            printf("  [enabled]  %s\n", name);
        else
            printf("  [disabled] %s\n", name);
    }
}

void pqc_kem_print_info(const PQC_KEM_Ctx *ctx) {
    if (!ctx) { printf("KEM context: NULL\n"); return; }
    printf("KEM algorithm     : %s\n",   ctx->algorithm);
    printf("  NIST level      : %u\n",   ctx->nist_level);
    printf("  IND-CCA2        : %s\n",   ctx->ind_cca ? "yes" : "no");
    printf("  Public-key len  : %zu B\n", ctx->length_public_key);
    printf("  Secret-key len  : %zu B\n", ctx->length_secret_key);
    printf("  Ciphertext len  : %zu B\n", ctx->length_ciphertext);
    printf("  Shared-secret   : %zu B\n", ctx->length_shared_secret);
}

/* ═══════════════════════════════════════════════════════════════════════════
 * SIG — context management
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_SIG_Ctx *pqc_sig_new(const char *alg) {
    const char *name = alg ? alg : PQC_SIG_DEFAULT;

    OQS_SIG *sig = OQS_SIG_new(name);
    if (!sig) {
        fprintf(stderr, "[pqc] SIG '%s': not available (unknown or disabled at build time)\n", name);
        return NULL;
    }

    PQC_SIG_Ctx *ctx = (PQC_SIG_Ctx *)calloc(1, sizeof(PQC_SIG_Ctx));
    if (!ctx) {
        OQS_SIG_free(sig);
        return NULL;
    }

    ctx->_sig               = sig;
    ctx->algorithm          = sig->method_name;
    ctx->length_public_key  = sig->length_public_key;
    ctx->length_secret_key  = sig->length_secret_key;
    ctx->length_signature   = sig->length_signature;
    ctx->nist_level         = (uint8_t)sig->claimed_nist_level;
    ctx->euf_cma            = sig->euf_cma;
    ctx->suf_cma            = sig->suf_cma;
    ctx->ctx_support        = OQS_SIG_supports_ctx_str(name);
    return ctx;
}

void pqc_sig_free(PQC_SIG_Ctx *ctx) {
    if (!ctx) return;
    OQS_SIG_free(ctx->_sig);
    ctx->_sig = NULL;
    free(ctx);
}

/* ═══════════════════════════════════════════════════════════════════════════
 * SIG — key-pair operations
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_STATUS pqc_sig_keygen(const PQC_SIG_Ctx *ctx, PQC_SIG_Keys *keys) {
    if (!ctx || !ctx->_sig || !keys) return PQC_ERR_NULL;

    memset(keys, 0, sizeof(*keys));
    keys->pk_len = ctx->length_public_key;
    keys->sk_len = ctx->length_secret_key;

    keys->public_key = (uint8_t *)malloc(keys->pk_len);
    keys->secret_key = (uint8_t *)malloc(keys->sk_len);
    if (!keys->public_key || !keys->secret_key) {
        pqc_sig_free_keys(keys);
        return PQC_ERR_ALLOC;
    }

    OQS_STATUS rc = OQS_SIG_keypair(ctx->_sig, keys->public_key, keys->secret_key);
    if (rc != OQS_SUCCESS) {
        pqc_sig_free_keys(keys);
        return PQC_ERR_KEYGEN;
    }
    return PQC_OK;
}

void pqc_sig_free_keys(PQC_SIG_Keys *keys) {
    if (!keys) return;
    _secure_free(&keys->public_key, keys->pk_len);
    _secure_free(&keys->secret_key, keys->sk_len);
    keys->pk_len = keys->sk_len = 0;
}

/* ═══════════════════════════════════════════════════════════════════════════
 * SIG — sign / verify
 * ═══════════════════════════════════════════════════════════════════════════ */

PQC_STATUS pqc_sig_sign(const PQC_SIG_Ctx *ctx,
                         const uint8_t     *message,
                         size_t             msg_len,
                         const uint8_t     *sk,
                         PQC_Signature     *sig_out) {
    if (!ctx || !ctx->_sig || !message || !sk || !sig_out) return PQC_ERR_NULL;

    memset(sig_out, 0, sizeof(*sig_out));
    sig_out->data = (uint8_t *)malloc(ctx->length_signature);
    if (!sig_out->data) return PQC_ERR_ALLOC;

    OQS_STATUS rc = OQS_SIG_sign(ctx->_sig,
                                  sig_out->data, &sig_out->len,
                                  message, msg_len,
                                  sk);
    if (rc != OQS_SUCCESS) {
        pqc_sig_free_signature(sig_out);
        return PQC_ERR_SIGN;
    }
    return PQC_OK;
}

PQC_STATUS pqc_sig_sign_ctx(const PQC_SIG_Ctx *ctx,
                              const uint8_t     *message,
                              size_t             msg_len,
                              const uint8_t     *ctx_str,
                              size_t             ctx_str_len,
                              const uint8_t     *sk,
                              PQC_Signature     *sig_out) {
    if (!ctx || !ctx->_sig || !message || !sk || !sig_out) return PQC_ERR_NULL;

    /* Fall back to plain sign when no context string supplied */
    if (!ctx_str || ctx_str_len == 0)
        return pqc_sig_sign(ctx, message, msg_len, sk, sig_out);

    if (!ctx->ctx_support) {
        fprintf(stderr, "[pqc] %s does not support context-string signing\n", ctx->algorithm);
        return PQC_ERR_GENERIC;
    }

    memset(sig_out, 0, sizeof(*sig_out));
    sig_out->data = (uint8_t *)malloc(ctx->length_signature);
    if (!sig_out->data) return PQC_ERR_ALLOC;

    OQS_STATUS rc = OQS_SIG_sign_with_ctx_str(ctx->_sig,
                                               sig_out->data, &sig_out->len,
                                               message, msg_len,
                                               ctx_str, ctx_str_len,
                                               sk);
    if (rc != OQS_SUCCESS) {
        pqc_sig_free_signature(sig_out);
        return PQC_ERR_SIGN;
    }
    return PQC_OK;
}

PQC_STATUS pqc_sig_verify(const PQC_SIG_Ctx *ctx,
                           const uint8_t     *message,
                           size_t             msg_len,
                           const uint8_t     *sig,
                           size_t             sig_len,
                           const uint8_t     *pk) {
    if (!ctx || !ctx->_sig || !message || !sig || !pk) return PQC_ERR_NULL;

    OQS_STATUS rc = OQS_SIG_verify(ctx->_sig, message, msg_len, sig, sig_len, pk);
    return (rc == OQS_SUCCESS) ? PQC_OK : PQC_ERR_VERIFY;
}

void pqc_sig_free_signature(PQC_Signature *sig) {
    if (!sig) return;
    if (sig->data) {
        OQS_MEM_cleanse(sig->data, sig->len);
        free(sig->data);
        sig->data = NULL;
    }
    sig->len = 0;
}

/* ═══════════════════════════════════════════════════════════════════════════
 * SIG — introspection utilities
 * ═══════════════════════════════════════════════════════════════════════════ */

void pqc_sig_list_algorithms(void) {
    printf("Available SIG algorithms:\n");
    for (size_t i = 0; i < OQS_SIG_algs_length; i++) {
        const char *name = OQS_SIG_alg_identifier(i);
        if (OQS_SIG_alg_is_enabled(name))
            printf("  [enabled]  %s\n", name);
        else
            printf("  [disabled] %s\n", name);
    }
}

void pqc_sig_print_info(const PQC_SIG_Ctx *ctx) {
    if (!ctx) { printf("SIG context: NULL\n"); return; }
    printf("SIG algorithm     : %s\n",   ctx->algorithm);
    printf("  NIST level      : %u\n",   ctx->nist_level);
    printf("  EUF-CMA         : %s\n",   ctx->euf_cma ? "yes" : "no");
    printf("  SUF-CMA         : %s\n",   ctx->suf_cma ? "yes" : "no");
    printf("  Context strings : %s\n",   ctx->ctx_support ? "yes" : "no");
    printf("  Public-key len  : %zu B\n", ctx->length_public_key);
    printf("  Secret-key len  : %zu B\n", ctx->length_secret_key);
    printf("  Max sig len     : %zu B\n", ctx->length_signature);
}
